Improve and fix some <soc.h> usages

gmarull opened the pull request:
<soc.h> has traditionally been used as a proxy to HAL headers,
register definitions, etc. Nowadays, <soc.h> is anarchy. It serves a
different purpose depending on the SoC. In some cases, it includes HALs,
in some others, it works as a header sink/proxy (for no good reason), as
a register definition when there's no HAL (bypassing the purpose of DT in some cases…).
To make things worse, it is being included in code that is, in theory, non-SoC specific.

This patch is part of a series intended to improve the situation by
removing <soc.h> usage when not needed, and by eventually removing it.
I've started mainly on arm64 as things are small enough to give it a try.

Some other small fixes have also been included, check commit messages.

Work related to https://github.com/zephyrproject-rtos/zephyr/issues/41543

gmarull committed:
arch: arm64: arm_mmu: make header self-contained

The header needs definitions for integer types and size_t, however, it
did not include the necessary headers. Note that this is only required
when not working with assembler.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
arch: arm64: mpu: remove unnecessary include

<soc.h> was not required.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: pl011: improve include list

<soc.h> was included because some CMSIS helpers (__DMB/__ISB) were
needed. In ARM SoCs, inclusion of CMSIS headers depends mainly on how
HALs decide to do it, being usually an inefficient and fragile include
chain. Note that on ARM64 we're in a better position, as those are
defined in-tree.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: arm64: fvp_aemv8a: fix soc.h usage

<soc.h> has been traditionally been used as a proxy to HAL headers,
register definitions, etc. Nowadays, <soc.h> is anarchy. It serves a
different purpose depending on the SoC. In some cases it includes HALs,
in some others it works as a header sink/proxy (for no good reason), as
a register definition when there's no HAL... To make things worse, it is
being included in code that is, in theory, non-SoC specific.

This patch is part of a series intended to improve the situation by
removing <soc.h> usage when not needed, and by eventually removing it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: arm64: fvp_aemv8r: remove empty soc.h

<soc.h> has been traditionally been used as a proxy to HAL headers,
register definitions, etc. Nowadays, <soc.h> is anarchy. It serves a
different purpose depending on the SoC. In some cases it includes HALs,
in some others it works as a header sink/proxy (for no good reason), as
a register definition when there's no HAL... To make things worse, it is
being included in code that is, in theory, non-SoC specific.

This patch is part of a series intended to improve the situation by
removing <soc.h> usage when not needed, and by eventually removing it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: arm64: nxp_imx: fix soc.h usage

<soc.h> has been traditionally been used as a proxy to HAL headers,
register definitions, etc. Nowadays, <soc.h> is anarchy. It serves a
different purpose depending on the SoC. In some cases it includes HALs,
in some others it works as a header sink/proxy (for no good reason), as
a register definition when there's no HAL... To make things worse, it is
being included in code that is, in theory, non-SoC specific.

This patch is part of a series intended to improve the situation by
removing <soc.h> usage when not needed, and by eventually removing it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: arm64: qemu_cortex_a53: fix soc.h usage

<soc.h> has been traditionally been used as a proxy to HAL headers,
register definitions, etc. Nowadays, <soc.h> is anarchy. It serves a
different purpose depending on the SoC. In some cases it includes HALs,
in some others it works as a header sink/proxy (for no good reason), as
a register definition when there's no HAL... To make things worse, it is
being included in code that is, in theory, non-SoC specific.

This patch is part of a series intended to improve the situation by
removing <soc.h> usage when not needed, and by eventually removing it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: arm64: bcm_vk: viper: improve soc.h include list

The header only required util_macro.h for BIT().

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: arm64: intel_socfpga: improve soc.h includes

arch/cpu.h was not needed, devicetree.h and sys/util.h are required for
DT API and ARRAY_SIZE().

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: arm64: nxp_layerscape: fix soc.h usage

<soc.h> has been traditionally been used as a proxy to HAL headers,
register definitions, etc. Nowadays, <soc.h> is anarchy. It serves a
different purpose depending on the SoC. In some cases it includes HALs,
in some others it works as a header sink/proxy (for no good reason), as
a register definition when there's no HAL... To make things worse, it is
being included in code that is, in theory, non-SoC specific.

This patch is part of a series intended to improve the situation by
removing <soc.h> usage when not needed, and by eventually removing it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: arm64: xenvm: fix soc.h usage

<soc.h> has been traditionally been used as a proxy to HAL headers,
register definitions, etc. Nowadays, <soc.h> is anarchy. It serves a
different purpose depending on the SoC. In some cases it includes HALs,
in some others it works as a header sink/proxy (for no good reason), as
a register definition when there's no HAL... To make things worse, it is
being included in code that is, in theory, non-SoC specific.

This patch is part of a series intended to improve the situation by
removing <soc.h> usage when not needed, and by eventually removing it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
arch: arm: aarch32: mpu: remove redundant soc.h usage

<soc.h> has been traditionally been used as a proxy to HAL headers,
register definitions, etc. Nowadays, <soc.h> is anarchy. It serves a
different purpose depending on the SoC. In some cases it includes HALs,
in some others it works as a header sink/proxy (for no good reason), as
a register definition when there's no HAL... To make things worse, it is
being included in code that is, in theory, non-SoC specific.

This patch is part of a series intended to improve the situation by
removing <soc.h> usage when not needed, and by eventually removing it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: arm64: intel_socfpga: remove dependency on soc.h

SOCFPGA_SYSMGR_REG_BASE is the only definition from soc.h used in tree,
move it to the file using it. Note that this looks suspicious: base
address should come from Devicetree. This patch will allow to drop
soc.h.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: arm64: intel_socfpga: delete redundant soc.h

<soc.h> has been traditionally been used as a proxy to HAL headers,
register definitions, etc. Nowadays, <soc.h> is anarchy. It serves a
different purpose depending on the SoC. In some cases it includes HALs,
in some others it works as a header sink/proxy (for no good reason), as
a register definition when there's no HAL... To make things worse, it is
being included in code that is, in theory, non-SoC specific.

This patch is part of a series intended to improve the situation by
removing <soc.h> usage when not needed, and by eventually removing it.

Note that in this case, none of the definitions are used in-tree.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: arm64: intel_socfpga: fix soc.h usage

<soc.h> has been traditionally been used as a proxy to HAL headers,
register definitions, etc. Nowadays, <soc.h> is anarchy. It serves a
different purpose depending on the SoC. In some cases it includes HALs,
in some others it works as a header sink/proxy (for no good reason), as
a register definition when there's no HAL... To make things worse, it is
being included in code that is, in theory, non-SoC specific.

This patch is part of a series intended to improve the situation by
removing <soc.h> usage when not needed, and by eventually removing it.

In this case, the file did not use anything defined in <soc.h>, it only
required access to integer types, e.g. uint32_t.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: ns16550: remove redundant soc.h include

<soc.h> has been traditionally been used as a proxy to HAL headers,
register definitions, etc. Nowadays, <soc.h> is anarchy. It serves a
different purpose depending on the SoC. In some cases it includes HALs,
in some others it works as a header sink/proxy (for no good reason), as
a register definition when there's no HAL... To make things worse, it is
being included in code that is, in theory, non-SoC specific.

This patch is part of a series intended to improve the situation by
removing <soc.h> usage when not needed, and by eventually removing it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: clock_control: agilex: remove redundant soc.h include

<soc.h> has been traditionally been used as a proxy to HAL headers,
register definitions, etc. Nowadays, <soc.h> is anarchy. It serves a
different purpose depending on the SoC. In some cases it includes HALs,
in some others it works as a header sink/proxy (for no good reason), as
a register definition when there's no HAL... To make things worse, it is
being included in code that is, in theory, non-SoC specific.

This patch is part of a series intended to improve the situation by
removing <soc.h> usage when not needed, and by eventually removing it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
arch: arm64: smp: remove redundant soc.h include

<soc.h> has been traditionally been used as a proxy to HAL headers,
register definitions, etc. Nowadays, <soc.h> is anarchy. It serves a
different purpose depending on the SoC. In some cases it includes HALs,
in some others it works as a header sink/proxy (for no good reason), as
a register definition when there's no HAL... To make things worse, it is
being included in code that is, in theory, non-SoC specific.

This patch is part of a series intended to improve the situation by
removing <soc.h> usage when not needed, and by eventually removing it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: arm64: fvp_aemv8r: improve include list

Make source self-contained by including what is needed.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: arm64: bcm_vk: viper: remove redundant soc.h usage

<soc.h> has been traditionally been used as a proxy to HAL headers,
register definitions, etc. Nowadays, <soc.h> is anarchy. It serves a
different purpose depending on the SoC. In some cases it includes HALs,
in some others it works as a header sink/proxy (for no good reason), as
a register definition when there's no HAL... To make things worse, it is
being included in code that is, in theory, non-SoC specific.

This patch is part of a series intended to improve the situation by
removing <soc.h> usage when not needed, and by eventually removing it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: arm64: qemu_cortex_a53: remove unused definition

SZ_1K was never used.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

povergoing commented:
That is to say, we will remove the `soc.h` eventually? I agreed that `soc.h` makes things worse. But can I ask that if you are sure there is always one or several methods to substitute all things that `soc.h` does? My concern might be if it turns out that we couldn't remove all the `soc.h`, someone who really needs `soc.h` will be confused about it, and sometimes it would be hard to decide whether to add it or not. Anyway, it's not a big problem, I agreed with this cleanup and hope there is little issue :)

povergoing reviewed (changes requested):
seems there is a build failure, https://github.com/zephyrproject-rtos/zephyr/runs/6675172044?check_suite_focus=true#step:12:3784

```
FAILED: zephyr/CMakeFiles/zephyr.dir/soc/arm64/nxp_imx/imx8m/dummy.c.obj 
ccache /opt/toolchains/zephyr-sdk-0.14.2/aarch64-zephyr-elf/bin/aarch64-zephyr-elf-gcc -DKERNEL -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -I../../../../../kernel/include -I../../../../../arch/arm64/include -I../../../../../include/zephyr -I../../../../../include -Izephyr/include/generated -I../../../../../soc/arm64/nxp_imx/imx8m -I../../../../../soc/arm64/nxp_imx/imx8m/. -I/__w/zephyr/modules/hal/atmel/include -isystem ../../../../../lib/libc/minimal/include -isystem /opt/toolchains/zephyr-sdk-0.14.2/aarch64-zephyr-elf/bin/../lib/gcc/aarch64-zephyr-elf/10.3.0/include -isystem /opt/toolchains/zephyr-sdk-0.14.2/aarch64-zephyr-elf/bin/../lib/gcc/aarch64-zephyr-elf/10.3.0/include-fixed -Os -imacros /__w/zephyr/zephyr/twister-out/imx8mp_evk_jailhouse/samples/synchronization/sample.kernel.synchronization/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-a53 -mabi=lp64 --sysroot=/opt/toolchains/zephyr-sdk-0.14.2/aarch64-zephyr-elf/aarch64-zephyr-elf -imacros /__w/zephyr/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -Werror -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/__w/zephyr/zephyr/samples/synchronization=CMAKE_SOURCE_DIR -fmacro-prefix-map=/__w/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/__w/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT zephyr/CMakeFiles/zephyr.dir/soc/arm64/nxp_imx/imx8m/dummy.c.obj -MF zephyr/CMakeFiles/zephyr.dir/soc/arm64/nxp_imx/imx8m/dummy.c.obj.d -o zephyr/CMakeFiles/zephyr.dir/soc/arm64/nxp_imx/imx8m/dummy.c.obj -c /__w/zephyr/zephyr/soc/arm64/nxp_imx/imx8m/dummy.c
/__w/zephyr/zephyr/soc/arm64/nxp_imx/imx8m/dummy.c:15:10: fatal error: soc.h: No such file or directory
   15 | #include <soc.h>
      |          ^~~~~~~
compilation terminated.
```

gmarull committed:
drivers: pm_cpu_ops: psci: remove redundant <soc.h> include

<soc.h> was not required.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: arm64: nxp_imx: remove dummy C file

A dummy C file was present in the imx8m SoC directory for no reason,
delete it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: arm64: nxp_layerscape: remove redundant soc.c source

The soc.c source file did nothing, so just remove it. Remove soc
directory from the include list as well, since there's no header in the
directory that needs to be exposed to the Zephyr build.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
> That is to say, we will remove the `soc.h` eventually? I agreed that `soc.h` makes things worse. But can I ask that if you are sure there is always one or several methods to substitute all things that `soc.h` does? My concern might be if it turns out that we couldn't remove all the `soc.h`, someone who really needs `soc.h` will be confused about it, and sometimes it would be hard to decide whether to add it or not. Anyway, it's not a big problem, I agreed with this cleanup and hope there is little issue :)

What I've started doing is to remove it from places where it is not necessary. I suspect this will end up with its removal, since there are better ways to do what `<soc.h>` does nowadays, but let's see.

> seems there is a build failure, [zephyrproject-rtos/zephyr/runs/6675172044?check_suite_focus=true#step:12:3784](https://github.com/zephyrproject-rtos/zephyr/runs/6675172044?check_suite_focus=true#step:12:3784)
> 
> ```
> FAILED: zephyr/CMakeFiles/zephyr.dir/soc/arm64/nxp_imx/imx8m/dummy.c.obj 
> ccache /opt/toolchains/zephyr-sdk-0.14.2/aarch64-zephyr-elf/bin/aarch64-zephyr-elf-gcc -DKERNEL -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -I../../../../../kernel/include -I../../../../../arch/arm64/include -I../../../../../include/zephyr -I../../../../../include -Izephyr/include/generated -I../../../../../soc/arm64/nxp_imx/imx8m -I../../../../../soc/arm64/nxp_imx/imx8m/. -I/__w/zephyr/modules/hal/atmel/include -isystem ../../../../../lib/libc/minimal/include -isystem /opt/toolchains/zephyr-sdk-0.14.2/aarch64-zephyr-elf/bin/../lib/gcc/aarch64-zephyr-elf/10.3.0/include -isystem /opt/toolchains/zephyr-sdk-0.14.2/aarch64-zephyr-elf/bin/../lib/gcc/aarch64-zephyr-elf/10.3.0/include-fixed -Os -imacros /__w/zephyr/zephyr/twister-out/imx8mp_evk_jailhouse/samples/synchronization/sample.kernel.synchronization/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-a53 -mabi=lp64 --sysroot=/opt/toolchains/zephyr-sdk-0.14.2/aarch64-zephyr-elf/aarch64-zephyr-elf -imacros /__w/zephyr/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -Werror -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/__w/zephyr/zephyr/samples/synchronization=CMAKE_SOURCE_DIR -fmacro-prefix-map=/__w/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/__w/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT zephyr/CMakeFiles/zephyr.dir/soc/arm64/nxp_imx/imx8m/dummy.c.obj -MF zephyr/CMakeFiles/zephyr.dir/soc/arm64/nxp_imx/imx8m/dummy.c.obj.d -o zephyr/CMakeFiles/zephyr.dir/soc/arm64/nxp_imx/imx8m/dummy.c.obj -c /__w/zephyr/zephyr/soc/arm64/nxp_imx/imx8m/dummy.c
> /__w/zephyr/zephyr/soc/arm64/nxp_imx/imx8m/dummy.c:15:10: fatal error: soc.h: No such file or directory
>    15 | #include <soc.h>
>       |          ^~~~~~~
> compilation terminated.
> ```

thanks, fixed (also fixed some other issues reported by CI).

carlocaione reviewed (dismissed):
no objections so far

povergoing commented:
> > That is to say, we will remove the `soc.h` eventually? I agreed that `soc.h` makes things worse. But can I ask that if you are sure there is always one or several methods to substitute all things that `soc.h` does? My concern might be if it turns out that we couldn't remove all the `soc.h`, someone who really needs `soc.h` will be confused about it, and sometimes it would be hard to decide whether to add it or not. Anyway, it's not a big problem, I agreed with this cleanup and hope there is little issue :)
> 
> What I've started doing is to remove it from places where it is not necessary. I suspect this will end up with its removal, since there are better ways to do what `<soc.h>` does nowadays, but let's see.

Fair enough :)

gmarull committed:
tests: kernel: context: fix soc.h include policy

Include only on platforms that are known to require it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

povergoing reviewed (dismissed):
refresh +1

gmarull committed:
arch: x86: core: early_serial: obtain NS16550 uart base address from DT

The NS16550 UART base address was hardcoded in <soc.h> headers. This
bypasses the console choice defined in Devicetree. Hardcoded hardware
choices must be avoided now that DT is in place.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: ns16550: use MMIO device depending on Kconfig option

Add a new selectable Kconfig option to decide wether the device driver
is a MMIO device or not. Previous to this patch, the decision was maded
based on the existence of a definition in <soc.h>. The design was
fragile, as code compiled anyway if the definition was not present.

All platforms/boards that had the definition in <soc.h> select the
Kconfig option in their respective defconfig files.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
soc: x86: remove unused UART_NS16550_ACCESS_IOPORT definition

The definition is no longer used, refer to previous commits for more
details.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

andyross commented on arch/x86/core/early_serial.c:
```diff
@@ -11,10 +11,10 @@
-#define IN(reg)       sys_in8(reg + UART_NS16550_ACCESS_IOPORT)
-#define OUT(reg, val) sys_out8(val, reg + UART_NS16550_ACCESS_IOPORT)
+#define IN(reg)       sys_in8(reg + DT_REG_ADDR(DT_CHOSEN(zephyr_console)))
+#define OUT(reg, val) sys_out8(val, reg + DT_REG_ADDR(DT_CHOSEN(zephyr_console)))
```
This code continues to go in circles.  The reason "early serial" doesn't use DTS is because this code tends to be brought up on new platforms before a working DTS configuration is present.

I've mostly given up on the idea that this can be maintained as a standalone driver that can be easily mutated during bringup, it's just too tempting to try to tie it into everything else.  So no -1, but...

andyross reviewed (approved):
Not loving the x86 serial changes, but not worth picking a fight over (basically: forcing the driver to use all the frameworks means that getting the frameworks working is harder, which IMHO is hurting and not helping).  But maybe now the EFI console option can fill that role and we can drop early serial.

Beyond that, totally agree on the principle here: soc.h doesn't actually define an API.  It's just a placeholder for "platform code can put junk into the global header environments here", which can be done a dozen other ways regardless.  There's no point in having it be a standard header at all, IMHO.
